### PR TITLE
Add C++17 build to Travis

### PR DESCRIPTION
**Pull request by sipa**
This adds a `--enable-c++17` option to the configure script, fixes the only C++17 incompatibility (with a commit taken from #18468), and adds a Travis test for it.

This is all off by default, and release builds remain C++11.

It implements the first step of the plan in https://github.com/bitcoin/bitcoin/issues/16684.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -61,8 +61,20 @@ case $host in
...
+
+AC_ARG_ENABLE([c++17],
+  [AS_HELP_STRING([--enable-c++17],
+  [enable compilation in c++17 mode (disabled by default)])],
+  [use_cxx17=$enableval],
+  [use_cxx17=no])
+
+dnl Require C++11 or C++17 compiler (no GNU extensions)
+if test "x$use_cxx17" = xyes; then
+  AX_CXX_COMPILE_STDCXX([17], [noext], [mandatory], [nodefault])
```
Comment (fanquake) [inline configure.ac:73]: Given it's been removed in the `ax_cxx_compile_stdcxx.m4` macro update, the `[nodefault]` argument should be removed here and below.

**Comment by fanquake** [comment]
Is this just to ensure continued C++17 compatibility of the codebase going forward?

**Comment by sipa** [comment]
@fanquake No, just the first step. See the plan in https://github.com/bitcoin/bitcoin/issues/16684.

**Comment by fanquake** [comment]
> No, just the first step. See the plan in #16684.

I mean, is the point of the CI addition just to ensure that after the last C++17 incompatibility is fixed (https://github.com/bitcoin/bitcoin/pull/18591/commits/10b4f256d4ab891e23977b1d3610a5b758f7bcb1), that there are no trivial regressions? I understand that this can't do much else.

**Comment by sipa** [comment]
Yeah, to demonstrate that it works, and to start introducing support in our build system.

File: configure.ac (hunk 1 above)
Comment (sipa) [reply configure.ac:73]: Done.

File: .travis.yml [hunk 2]
```diff
@@ -148,6 +148,11 @@ jobs:
       env: >-
         FILE_ENV="./ci/test/00_setup_env_native_valgrind.sh"
 
+    - stage: test
+      name: 'x86_64 Linux  [GOAL: install]  [focal]  [no depends, only system libs, c++17]'
```
Comment (MarcoFalke) [inline .travis.yml:152]: Our releases are built with bionic, and I think there are no plans to drop support for bionic in 0.22.0. So this should probably run on bionic?

**Review by maflcko — COMMENTED** [review]
Concept ACK

File: .travis.yml (hunk 2 above)
Comment (sipa) [reply .travis.yml]: Done. Travis passes.

**Comment by hebasto** [comment]
Concept ACK.

File: configure.ac (hunk 1 above)
Comment (hebasto) [reply configure.ac]: The `[nodefault]` argument was added in _our_ patch: https://github.com/bitcoin/bitcoin/pull/9831/commits/9829c54de2725037ee0702306cbaa99fc9aa1826

Why is it possible to drop it?

Comment (sipa) [reply configure.ac]: Ouch. I had no idea this was a local modification.

Comment (sipa) [reply configure.ac]: Ah, it seems the new upstream version actually changed behaviour and now does what we called "nodefault" regardless.

Even with compilers that support c++17, if we request c++11, we'll only get c++11 (which can be verified from the Travis build logs).

Comment (fanquake) [reply configure.ac]: I should have made that more clear. I’ve got the same change as part of #18088.

**Comment by elichai** [comment]
Concept ACK

For ease review this is the same implicit conversion impl that libc++ has: https://github.com/llvm-mirror/libcxx/blob/master/include/span#L229

**Review by hebasto — APPROVED** [review]
ACK c31cbe7cfefc18123eb85ffb2ce509748435efde, tested on Linux Mint 19.3 both C++11 and C++17 modes. Compiled and passed tests locally.

Verified, that build fails in C++17 mode without "Support conversion between Spans of compatible types" commit (0fbde488b24f62b4bbbde216647941dcac65c81a).

**Comment by practicalswift** [comment]
Concept ACK

Asserting that we compile and pass tests also with a C++17 mode build is valuable.

**Comment by laanwj** [comment]
Concept ACK, thanks for adding.

**Comment by elichai** [comment]
tACK c31cbe7cfefc18123eb85ffb2ce509748435efde

Checked that with and without `--enable-c++17` it compiles(including Qt and everything) and tests pass.
Checked with `make -n` that without the flag it still passes `-std=c++11` and with it it passes `-std=c++17`.
Checked that with the span changes reverted it still works on c++11 but fails on c++17 with:
<details>

```
test/util_tests.cpp: In member function ‘void util_tests::test_spanparsing::test_method()’:                                                                                                                                                                                                                                                                                                  
test/util_tests.cpp:1832:24: error: no match for ‘operator=’ (operand types are ‘Span<const char>’ and ‘Span<char>’)                                                                                                                                                                                                                                                                         
 1832 |     sp = MakeSpan(input);                                                                                                                                                                                                                                                                                                                                                            
      |                        ^                                                                                                                                                                                                                                                                                                                                                             
In file included from ./serialize.h:25,                                                                                                                                                                                                                                                                                                                                                      
                 from ./hash.h:13,                                                                                                                                                                                                                                                                                                                                                           
                 from test/util_tests.cpp:8:                                                                                                                                                                                                                                                                                                                                                 
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(const Span<const char>&)’                                                                                                                                                                                                                                                                           
   18 | class Span                                                                                                                                                                                                                                                                                                                                                                           
      |       ^~~~                                                                                                                                                                                                                                                                                                                                                                           
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘const Span<const char>&’                                                                                                                                                                                                                                                                                     
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(Span<const char>&&)’                                                                                                                                                                                                                                                                                
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘Span<const char>&&’                                                                                                                                                                                                                                                                                          
test/util_tests.cpp:1857:24: error: no match for ‘operator=’ (operand types are ‘Span<const char>’ and ‘Span<char>’)                                                                                                                                                                                                                                                                         
 1857 |     sp = MakeSpan(input);                                                                                                                                                                                                                                                                                                                                                            
      |                        ^                                                                                                                                                                                                                                                                                                                                                             
In file included from ./serialize.h:25,                                                                                                                                                                                                                                                                                                                                                      
                 from ./hash.h:13,                                                                                                                                                                                                                                                                                                                                                           
                 from test/util_tests.cpp:8:                                                                                                                                                                                                                                                                                                                                                 
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(const Span<const char>&)’                                                                                                                                                                                                                                                                           
   18 | class Span                                                                                                                                                                                                                                                                                                                                                                           
      |       ^~~~                                                                                                                                                                                                                                                                                                                                                                           
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘const Span<const char>&’
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(Span<const char>&&)’
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘Span<const char>&&’
test/util_tests.cpp:1880:24: error: no match for ‘operator=’ (operand types are ‘Span<const char>’ and ‘Span<char>’)
 1880 |     sp = MakeSpan(input);
      |                        ^
In file included from ./serialize.h:25,
                 from ./hash.h:13,
                 from test/util_tests.cpp:8:
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(const Span<const char>&)’
   18 | class Span
      |       ^~~~
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘const Span<const char>&’
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(Span<const char>&&)’
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘Span<const char>&&’
test/util_tests.cpp:1886:24: error: no match for ‘operator=’ (operand types are ‘Span<const char>’ and ‘Span<char>’)
 1886 |     sp = MakeSpan(input);
      |                        ^
In file included from ./serialize.h:25,
                 from ./hash.h:13,
                 from test/util_tests.cpp:8:
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(const Span<const char>&)’
   18 | class Span
      |       ^~~~
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘const Span<const char>&’
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(Span<const char>&&)’
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘Span<const char>&&’
test/util_tests.cpp:1892:24: error: no match for ‘operator=’ (operand types are ‘Span<const char>’ and ‘Span<char>’)
 1892 |     sp = MakeSpan(input);
      |                        ^
In file included from ./serialize.h:25,
                 from ./hash.h:13,
                 from test/util_tests.cpp:8:
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(const Span<const char>&)’
   18 | class Span
      |       ^~~~
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘const Span<const char>&’
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(Span<const char>&&)’
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘Span<const char>&&’
test/util_tests.cpp:1898:24: error: no match for ‘operator=’ (operand types are ‘Span<const char>’ and ‘Span<char>’)
 1898 |     sp = MakeSpan(input);
      |                        ^
In file included from ./serialize.h:25,
                 from ./hash.h:13,
                 from test/util_tests.cpp:8:
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(const Span<const char>&)’
   18 | class Span
      |       ^~~~
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘const Span<const char>&’
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(Span<const char>&&)’
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘Span<const char>&&’
test/util_tests.cpp:1904:24: error: no match for ‘operator=’ (operand types are ‘Span<const char>’ and ‘Span<char>’)
 1904 |     sp = MakeSpan(input);
      |                        ^
In file included from ./serialize.h:25,
                 from ./hash.h:13,
                 from test/util_tests.cpp:8:
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(const Span<const char>&)’
   18 | class Span
      |       ^~~~
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘const Span<const char>&’
./span.h:18:7: note: candidate: ‘constexpr Span<const char>& Span<const char>::operator=(Span<const char>&&)’
./span.h:18:7: note:   no known conversion for argument 1 from ‘Span<char>’ to ‘Span<const char>&&’
test/util_tests.cpp:1913:29: error: invalid initialization of reference of type ‘const Span<const char>&’ from expression of type ‘Span<char>’
 1913 |     results = Split(MakeSpan(input), 'x');
      |                     ~~~~~~~~^~~~~~~
In file included from test/util_tests.cpp:20:
./util/spanparsing.h:46:61: note: in passing argument 1 of ‘std::vector<Span<const char> > spanparsing::Split(const Span<const char>&, char)’
   46 | std::vector<Span<const char>> Split(const Span<const char>& sp, char sep);
      |                                     ~~~~~~~~~~~~~~~~~~~~~~~~^~
test/util_tests.cpp:1921:29: error: invalid initialization of reference of type ‘const Span<const char>&’ from expression of type ‘Span<char>’
 1921 |     results = Split(MakeSpan(input), '-');
      |                     ~~~~~~~~^~~~~~~
In file included from test/util_tests.cpp:20:
./util/spanparsing.h:46:61: note: in passing argument 1 of ‘std::vector<Span<const char> > spanparsing::Split(const Span<const char>&, char)’
   46 | std::vector<Span<const char>> Split(const Span<const char>& sp, char sep);
      |                                     ~~~~~~~~~~~~~~~~~~~~~~~~^~
test/util_tests.cpp:1926:29: error: invalid initialization of reference of type ‘const Span<const char>&’ from expression of type ‘Span<char>’
 1926 |     results = Split(MakeSpan(input), '#');
      |                     ~~~~~~~~^~~~~~~
In file included from test/util_tests.cpp:20:
./util/spanparsing.h:46:61: note: in passing argument 1 of ‘std::vector<Span<const char> > spanparsing::Split(const Span<const char>&, char)’
   46 | std::vector<Span<const char>> Split(const Span<const char>& sp, char sep);
      |                                     ~~~~~~~~~~~~~~~~~~~~~~~~^~
test/util_tests.cpp:1933:29: error: invalid initialization of reference of type ‘const Span<const char>&’ from expression of type ‘Span<char>’
 1933 |     results = Split(MakeSpan(input), '*');
      |                     ~~~~~~~~^~~~~~~
In file included from test/util_tests.cpp:20:
./util/spanparsing.h:46:61: note: in passing argument 1 of ‘std::vector<Span<const char> > spanparsing::Split(const Span<const char>&, char)’
   46 | std::vector<Span<const char>> Split(const Span<const char>& sp, char sep);
```

</details>

**Comment by practicalswift** [comment]
Tested ACK c31cbe7cfefc18123eb85ffb2ce509748435efde

**Pull request merged**